src/lake/Lake/Load/Elab.lean

```diff
@@ -65,3 +76,30 @@ def elabConfigFile (pkgDir : FilePath) (configOpts : NameMap String)
     error s!"{configFile}: package configuration has errors"
   else
     return s.commandState.env
+
+/--
+If `reconfigure` is not set and up-to-date OLean for the configuration file exists,
+import it. Otherwise, elaborate the configuration and store save it to the OLean.
+-/
+def importConfigFile (pkgDir : FilePath) (lakeOpts : NameMap String)
+(leanOpts := Options.empty) (configFile := pkgDir / defaultConfigFile) (reconfigure := true) : LogIO Environment := do
+  let olean := configFile.withExtension "olean"
+  let useOLean ← id do
+    if reconfigure then return false
+    unless (← olean.pathExists) do return false
+    unless (← getMTime olean) > (← getMTime configFile) do return false
+    return true
+  if useOLean then
+    withImporting do
+    let (mod, region) ← readModuleData olean
+    let s ← importModulesUsingCache mod.imports
+    let s := {s with
+      moduleData  := s.moduleData.push mod
+      regions     := s.regions.push region
+      moduleNames := s.moduleNames.push configModuleName
+    }
+    finalizeImport s #[configModuleName] leanOpts 1024
+  else
+    let env ← elabConfigFile pkgDir lakeOpts leanOpts configFile
+    Lean.writeModule env olean
```

Comment:
I don't like that this is putting `lakefile.olean`, a build file, outside the `build` directory. My suggestion is to put it in `build/lakefile.olean`.

Reply:
@digama0 I am incline to think of `lakefile.olean` as more of static lock file like `lake-manifest.json` than a build file. The only major difference is that `lakefile.olean` is a platform/configuration-specific lock, hence its exclusion from Git. Also, a technical problem with putting it in the build directory is that Lake currently does not know *what the build directory is* until we load the configuration file.

Reply:
@digama0 When you say it should be supplied "by the user", do you mean a CLI option? Note that, in addition tot he environment variable, Cargo also allows you to specify the build directory via the configuration file with the [`build.target-dir`](https://doc.rust-lang.org/cargo/reference/config.html#buildtarget-dir) setting.

Reply:
A CLI option would be fine, as would a config file (but that would be a significant new feature, here "config file" means a global or project-local setting which would generally not be committed). The environment variable is the lowest effort version of this which still gives the basic feature without all the issues involved in tying it up in lean elaboration.

Reply:
@digama0 For Lake, the elaborated configuration *is* the project-local "config file". I do not envision the creation of any other. If it could be reasonable stuck in a such a thing, it should be part of the lakefile then (as it currently is). A standard CLI override does sound like a good idea for a future feature though!

Reply:
One caveat to my statement about the creation of another static config file: I can imagine creating declarative config file after an elaboration that contains some subset of the olean data that can be declaratively represented.

Reply:
The "config file" here is for things specific to the user's machine. It would normally just be global and in the user's home directory, but cargo also allows you to customize this on a per-subfolder (including per-project) basis. This is a very different thing from the lakefile / Cargo.toml, which is about properties of the project itself, and lake currently has almost no need for such a thing.